### ISSUE TITLE
Environment object - Damaging thing

### ISSUE THREAD

**Issue by perfecthumaninterface** [alkaitagi/ProjectBurst#50]
A new environment object such as spikes or even just a glowing orange block that will damage units that come in contact with it. 

It should not be a "structure;" it has no health and is a static part of the level. 

When units or the player collide with it, damage is dealt, and a physics impulse is applied to knock them away.

**Comment by alkaitagi** [comment]
Added **Explosion** component. Leave _timer_ at 0 or less to make it persistent.

![image](https://user-images.githubusercontent.com/32439229/71618082-28be3480-2bcf-11ea-9f77-4760c151af8d.png)

**Comment by perfecthumaninterface** [comment]
Was this also designed to be used elsewhere, such as spawned with bullets and such?

**Comment by alkaitagi** [comment]
No, bullets utilize different implementation - projectile modifiers, although you could spawn this as well, but that's unnecessary.

**Comment by perfecthumaninterface** [comment]
Ok, just trying to understand the thought process behind theming it as an "explosion." I had envisioned it as just a sort of block to be placed in the level. Of course we can change the collider and visual element easily enough. Also I'm curious about the "Timer," what that's intended for. Were you thinking to use it like a trap to dodge through like in Sharp?

Unit interaction with the object is a bit janky. I think the fix is to halt all other movement (the acceleration the unit is trying to do on its own) of the unit for a time after being damaged by it. I can only guess the multiple forces being applied is messing with the interpolation or something. Maybe you have a better idea.

Also, if you're not doing it already, I think it would be a good idea to zero-out the velocity of the unit immediately before applying the impulse, to keep the action consistent. I don't think this is any issue currently but it may prevent future issues.

**Comment by alkaitagi** [comment]
To be honest I just named script after the prefab it was going to be attached to, so I did not really put much thoughts into it. Of course current naming is confusing, therefore renamed it as __DamageObject__.

Timer property is just something I thought you might need, but now I see it as being redundant so that's removed as well.

Your suggestion with setting rigidbody's velocity to 0 prior to applying impulse is completely sound. Implementing that (for projectile's impulse modifier as well) has improved the object's interaction with moving units.

**Comment by perfecthumaninterface** [comment]
Unfortunately not yet perfect. If you set the impulse to a higher value like 100 then you can see some jerky motion. I expect my previous suggestion will probably work but it might be nice to understand why it does that. 

On a separate issue, we should make this into a pathfinding obstacle. But with things currently it would be difficult to test the object if we do that, so maybe hold off until we're satisfied with the other aspects of it.

**Comment by alkaitagi** [comment]
Please, evaluate the physics again. I've added rigidbody's disable when a Unit is pushed by impulse, so neither player's nor AI's input is accepted by movement system.

**Comment by perfecthumaninterface** [comment]
I think it's different now but there is still some odd roughness to the movement when creeps bounce off of it. Also the motion seems a bit awkward when an enemy receives an impulse but then collides with other enemies. 

I'm not sure what to suggest. I wonder if using custom movement code instead of physics for some of these types of things may be better. It seems like trying to do unnatural movement within physics is difficult. But I don't know.
